tests-cuda/test_1276-cuda-num.py

```diff
@@ -127,7 +127,8 @@ def test_num_8():
     )
 
     cuda_recordarray = ak.to_backend(recordarray, "cuda")
-    assert ak.num(cuda_recordarray, 0).tolist() == ak.num(recordarray, 0).tolist()
+
```

Comment:
@agoose77 - this is unrelated to CuPy. It looks like the `ak.num` returns a scalar, not a 0-D array now.

Reply:
The current policy is that Awkward's scalar outputs have NumPy scalar types (or more specifically, whatever the scalar output of the underlying array library is; for CuPy, that's zero-dimensional arrays). In other words, we do as little as possible to the array library's output—just pass it through.

Wrapping CuPy scalars (zero-dimensional arrays) with `int` is a synchronization point. In general, CUDA kernels run asynchronously and Python only waits for a kernel output if it needs it for something, such as printing it to the screen. Converting a ([promised](https://en.wikipedia.org/wiki/Futures_and_promises)) CuPy zero-dimensional array into a Python `int` will make Python sit and wait for all of the CUDA processes that array depends on to finish.

Is it not the case that

```python
self._content[startsmin : self._stops.data.max()]
```

can accept any `numbers.Integral` number, not necessarily `int`? That's why we try to keep our integer-checking generic like that. I think there's an `ak._util.is_int` helper function to check for "any integer type but not boolean," so that it will accept NumPy's `np.int64` or CuPy's `cp.ndarray` with `shape=()`.

I see in the original issue, the error message was

```
TypeError: slice indices must be integers or None or have an __index__ method
```

I don't see that error message in our codebase. Is it Awkward complaining or CuPy? Does it pass our `is_int` check but fail one of CuPy's? (Is it really the case that CuPy returns values from `max` that can't be used as slice endpoints? It seems to be accepting `array[:array.max()]` where `array` is a CuPy array.)

I'm wondering if this is the right place/way to fix this bug.

Reply:
I see; CuPy wants us to know that we're slowing down the GPU. But in this case, we have to, so this is the best solution to the problem.

Both `x.item()` and `int(x)` work, but I prefer `x.item()`, too.

Reply:
I wonder if we should introduce `ak.Index.max()` and `ak.Index.min()` so that instead of `int(self._stops.data.max())` we'd have `self._stops.max()`

Reply:
@agoose77 and @jpivarski - Please, let me know if I should revert it to `x.item()`. Otherwise, please merge it. Thanks!

Reply:
I'd prefer to not have `Index.max` and `Index.min`, since the same argument could be made for all of an array's methods, and that's a lot. Also, @agoose77 is planning to remove method-calls in favor of nplike functions, so that would be going away, anyway.

I have only a slight preference for `item`. But I can do it in the web interface, accept my own changes, and turn on auto-merge if you're done.

Thanks!

Reply:
Thanks!